cachi2/core/package_managers/yarn/project.py

```diff
@@ -47,6 +56,111 @@ def cache_folder(self) -> str:
         """
         return self._data.get("cacheFolder", DEFAULT_CACHE_FOLDER)
 
+    @property
+    def checksum_behavior(self) -> Optional[ChecksumBehavior]:
```

Comment:
The reasoning was to exactly limit the amount of data we could pull from this file, and to provide a clear, high-level interface to access this data (that's why we have methods such as `registry_server_for_scope`, which does not translate directly to a yarnrc property).

I completely agree that it does not scale, and I was counting on the "we won't use most of them" part when this was designed. I need to say I did get annoyed by adding these new properties here in this PR.

I'm not against the idea of completely dumping this approach to basically implementing a config map class, but we'd need to do it for both YarnRc and PackageJson, otherwise we'd lose consistency. Maybe also revisit the purpose of the Project class.

Reply:
The idea about adding defaults is something we might not want to do, since we also use this class to write the file to disk, and we'd end up writing a much larger file than it's actually necessary. That's the reason why most properties return `None` even if a default value exists.

I know we kind of reached a mixed approach because we have the `DEFAULT_YARN_` constants, but I think it's better to dump those and essentially return what is defined in the file (`None` if the config key is not set) than to set every default there is.

Reply:
Now, all in all, I'd say this should be handled in a separate PR. I'd say let's move this discussion to #377.

Reply:
> The reasoning was to exactly limit the amount of data we could pull from this file, and to provide a clear, high-level interface to access this data (that's why we have methods such as registry_server_for_scope, which does not translate directly to a yarnrc property). 

I think this largely depends on who the intended audience and intended use cases are for such an interface. Since we're not functioning as a library and we're abstracting this for our own internal use case I don't see a reason to limit the scope and just create a full-fledged config map (I would argue similarly even if it were a user interface, since IMO these limited interfaces make sense mainly for REST API wrappers, not configurations).

> I'm not against the idea of completely dumping this approach to basically implementing a config map class, but we'd need to do it for both YarnRc and PackageJson, otherwise we'd lose consistency. Maybe also revisit the purpose of the Project class.

Uhm, sure, I focused solely on `YarnRc` and not Package.json in #377 because that's what I bumped into as I needed field/s that weren't exposed and as I was working on that you proposed this so my hunch that we'd need more fields was correct. I agree that consistency is important, but it doesn't mean both need to be converted at the same time though (that's a separate discussion 

> The idea about adding defaults is something we might not want to do, since we also use this class to write the file to disk, and we'd end up writing a much larger file than it's actually necessary. That's the reason why most properties return None even if a default value exists.

Even though #377  doesn't handle ^this at the moment, I think this is essentially a non-issue, because there's nothing stopping us to implement the config map in a way where defaults are stored in a separate structure, merging the values on demand (I already expected this argument during my work on the mentioned PR), so we'd never dump a massive Rc file to the disk.

> I know we kind of reached a mixed approach because we have the DEFAULT_YARN_ constants, but I think it's better to dump those and essentially return what is defined in the file (None if the config key is not set) than to set every default there is.

See my argument above.

> Now, all in all, I'd say this should be handled in a separate PR. Let's move this discussion to https://github.com/containerbuildsystem/cachi2/pull/377.

Sure. I wasn't trying to stop this PR, I was just genuinely curious as my PR very clearly conflicts with this approach and I wanted to understand the reasoning to figure out whether #377 needs to be thrown to the bin or not.

Reply:
> I think this largely depends on who the intended audience and intended use cases are for such an interface.
I'd probably stick with the current approach until we need to add more attributes, since it is not a real problem yet, and we likely covered all we need already. But, you already invested work into the new approach, so I just want to make sure we agree it's what we want, and we proceed with #277 and already plan moving PackageJson (in a different PR, not in #377). 

Let's see what other folks think. We can probably start a thread in #377 to decide.

Reply:
> I think this is essentially a non-issue, because there's nothing stopping us to implement the config map in a way where defaults are stored in a separate structure

It is added complexity, as even loading the defaults is. My question here would be if we really need the defaults at all. Maybe the only one we actually need is the default yarn cache path, and we could simply return `None` for any config that is missing from the config map. 

I'd stick with the simple approach unless we really need to go complex.

Reply:
I wouldn't have opened this discussion if I didn't make use of another config option already  - `nodeLinker`. At the time of writing my patches there were only a few options covered and so I went down the rabbit hole. At this point I'd say we never know how many we'll need so instead of just blindly adding  more properties, loading the defaults and be done with it for good to me sounds like a better long term plan. However, I stick to my original comment that we don't need to change anything in context of this specific PR as that would be too much work, that is what #377 should cover.

Reply:
I'll close this, and we can move this discussion to #377 then.